Sprint 12

Georgy-hook opened the pull request.

Georgy-hook committed:
- created a network layer for ImageListViewController

Georgy-hook committed:
- ImageListCellDelegate added
- Changed DateFormatter
- ImageListViewController present images from API now
- Added the ability to exit the profile
- SingleImageViewController present images from API now
- Changed ErrorAlertService

helecta commented on ImageFeed/Helpers/URlSession + URLRequest.swift:
```diff
@@ -46,6 +46,7 @@ extension URLSession {
                         let result = try decoder.decode(T.self, from: data)
                         fulfillCompletionOnMainThread(.success(result))
                     } catch {
+                        print(error)
```
обычно перед сдачей принято удалять закомментированный код и print’ы. Так код проще читать, а дебажная информация не должна попадать в продакшен(т.е print’ы)

helecta commented on ImageFeed/Presentation/ImagesListViewController.swift:
```diff
@@ -18,14 +18,18 @@ final class ImagesListViewController: UIViewController {
-    private let ShowSingleImageSegueIdentifier = "ShowSingleImage"
+    var photos: [Photo] = []
+    private let imagesListService = ImagesListService.shared
+    private var ImagesListServiceObserver: NSObjectProtocol?
```
такие переменные стоит называть со строчной буквы. С заглавной можно называть только типы и протоколы.
Почитать подробнее можно  [тут](https://github.com/RedMadRobot/RMR-swift-style-guide#naming)  или  [тут](https://github.com/tutu-ru/swift-style-guide#%D0%B8%D0%BC%D0%B5%D0%BD%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5)

helecta commented on ImageFeed/Presentation/ImagesListViewController.swift:
```diff
@@ -18,14 +18,18 @@ final class ImagesListViewController: UIViewController {
         addSubviews()
         applyConstraints()
+        addObserver()
+        imagesListService.fetchPhotosNextPage(OAuth2TokenStorage.shared.token!)
```
Лучше переписать без force unwrapping. Можно использовать guard. Force unwrapping это всегда потенциальное место, где приложение в будущем может упасть.
Поэтому стоит взять себе за привычку если можно без force unwrapping написать, то нужно писать без.

helecta commented on ImageFeed/Presentation/ImagesListViewController.swift:
```diff
@@ -56,35 +55,34 @@ extension ImagesListViewController:UITableViewDataSource{
         present(singleImageViewController, animated: true)
     }
+    func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
+        imagesListService.fetchPhotosNextPage(OAuth2TokenStorage.shared.token!)
```
И тут тоже force unwrapping.

helecta commented on ImageFeed/Presentation/ImagesListViewController.swift:
```diff
@@ -108,3 +106,55 @@ extension ImagesListViewController{
+            switch result {
+            case .success:
+
+                self.photos = self.imagesListService.photos
```
В замыкание для избежания retain cycl'ов, необходимо передавать слабую ссылку на `self`

helecta commented on ImageFeed/Presentation/SingleImageViewController.swift:
```diff
@@ -42,23 +42,32 @@ final class SingleImageViewController: UIViewController {
+            imageView.kf.indicatorType = .activity
+            imageView.kf.setImage(with: url,
+                                  completionHandler: {_ in
+                self.rescaleAndCenterImageInScrollView(image: self.imageView.image!)
```
Строгая ссылка на self. Лучше в захвате замыкания сделать слабую

helecta replied:
`self.imageView.image!` - force unwrapping, можно переписать без него

helecta commented on ImageFeed/Presentation/SingleImageViewController.swift:
```diff
@@ -42,23 +42,32 @@ final class SingleImageViewController: UIViewController {
-            guard let image = image else { return }
-            imageView.image = image
-            rescaleAndCenterImageInScrollView(image: image)
+            guard let url = URL(string: imageURL!) else { return }
```
`imageURL!` - force unwrapping, можно переписать без него

helecta commented on ImageFeed/Presentation/SingleImageViewController.swift:
```diff
@@ -42,23 +42,32 @@ final class SingleImageViewController: UIViewController {
-        guard let image = image else { return }
-        imageView.image = image
-        rescaleAndCenterImageInScrollView(image: image)
+        guard let url = URL(string: imageURL!) else { return }
```
imageURL! - force unwrapping, можно переписать без него

helecta commented on ImageFeed/Presentation/SingleImageViewController.swift:
```diff
@@ -42,23 +42,32 @@ final class SingleImageViewController: UIViewController {
+        imageView.kf.indicatorType = .activity
+        imageView.kf.setImage(with: url,
+                              completionHandler: {_ in
+            self.rescaleAndCenterImageInScrollView(image: self.imageView.image!)
```
Строгая ссылка на self. Лучше в захвате замыкания сделать слабую

self.imageView.image! - force unwrapping, можно переписать без него

helecta commented on ImageFeed/Services/ImagesListService.swift:
```diff
@@ -0,0 +1,137 @@
+    
+    private var lastLoadedPage: Int?
+    
+    static let DidChangeNotification = Notification.Name(rawValue: "ImagesListProviderDidChange")
```
Такие константы стоит называть со строчной буквы.  Да, в учебнике было так, но все-таки правильно называть со строчной. С заглавной можно называть только типы и протоколы.
Почитать подробнее можно  [тут](https://github.com/RedMadRobot/RMR-swift-style-guide#naming)  или  [тут](https://github.com/tutu-ru/swift-style-guide#%D0%B8%D0%BC%D0%B5%D0%BD%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5)

helecta commented on ImageFeed/Services/ImagesListService.swift:
```diff
@@ -0,0 +1,137 @@
+        
+        let nextPage = lastLoadedPage == nil
+        ? 1
+        : lastLoadedPage! + 1
```
Тут можно переписать без force unwrap.
Force операции могут приводить к падению приложения, считается что падение приложения хуже, чем не рабочее состояние. Да, тут падение невозможно, но стоит взять себе за привычку если можно без force unwrap написать, то нужно писать без.

Как вариант можно сделать дефолтное значение lastLoadedPage:
`var lastLoadedPage: Int = 1`

helecta commented on ImageFeed/Services/ImagesListService.swift:
```diff
@@ -0,0 +1,137 @@
+            request = ImagesLikeURLRequest(authToken: OAuth2TokenStorage.shared.token!, id: photoId)
+        }
+        let session = URLSession.shared
+        let task = session.dataTask(with: request!) {[weak self] data, response, error in
```
`request!` - force unwrapping, можно без него переписать

helecta commented on ImageFeed/Services/ImagesListService.swift:
```diff
@@ -0,0 +1,137 @@
+            request = ImagesDeleteLikeURLRequest(authToken: OAuth2TokenStorage.shared.token!, id: photoId)
+        }
+        else {
+            request = ImagesLikeURLRequest(authToken: OAuth2TokenStorage.shared.token!, id: photoId)
```
`OAuth2TokenStorage.shared.token!` - force unwrapping, можно без него переписать

helecta commented on ImageFeed/Services/ImagesListService.swift:
```diff
@@ -0,0 +1,137 @@
+    func changeLike(photoId: String, isLike: Bool, _ completion: @escaping (Result<Void, Error>) -> Void){
+        let request: URLRequest?
+        if isLike {
+            request = ImagesDeleteLikeURLRequest(authToken: OAuth2TokenStorage.shared.token!, id: photoId)
```
`OAuth2TokenStorage.shared.token!` - force unwrapping, можно без него переписать

helecta commented on ImageFeed/Helpers/DateFormatter.swift:
```diff
@@ -8,10 +8,19 @@
 private var dateFormatter: DateFormatter = {
     let formatter = DateFormatter()
-    formatter.dateStyle = .long
+    formatter.dateStyle = .medium
```
Если так настроить дата форматтер:
```
private lazy var dateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "dd MMMM yyyy"
        formatter.locale = Locale(identifier: "ru_RU")
        return formatter
    }()
```
То будет выглядеть как в макете, без  `г.`

helecta commented on ImageFeed/Helpers/DateFormatter.swift:
```diff
@@ -8,10 +8,19 @@
 }
+extension String{
+    func getDate() -> Date? {
+        let dateFormatter = DateFormatter()
```
форматер будет создаваться каждый раз при вызове метода, а по заданию нужно сделать чтобы все форматеры создавались только один раз

helecta commented on ImageFeed/Presentation/SplashScreenViewController.swift:
```diff
@@ -48,7 +46,7 @@ final class SplashScreenViewController: UIViewController {
     }
     
-
+    
```
Сейчас у тебя такой поток авторизации:
1. Приложение запускается, показан сплеш экран
2. Если пользователь не авторизован, показывается экран авторизации
3. Пользователь авторизовываются во вебвью
4. Вебвью закрывается и показывается сплеш скрин с лоудингом
5. Успевает показаться еще раз экран авторизации перед тем как будет показана лента

Это происходит потому что в SplashScreenViewController, когда в методе `func authViewController(_ vc: AuthViewController, didAuthenticateWithCode code: String)` делаешь dismiss, то повторно вызывается метод viewDidAppear, а т.к токен еще не успели получить к этому моменту, то будет проскакивать экран авторищзации. 

Причем еще будет проблема и с показом алерта, если произошла ошибка. Алерт не успеет показаться, т.к у тебя сразу будет вызван метод `viewDidAppear`, в котором сразу будет показан экран авторизации.

Как вариант решения этих проблем:
1. Добавить в UIBlockingProgressHUD свойство `isShowing`:
```
final class UIBlockingProgressHUD {
    
    static var isShowing: Bool = false
    
    private static var window: UIWindow? {
        guard let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
              let window = windowScene.windows.first
        else {
            return nil
        }
        return window
    }

    static func show() {
        isShowing = true
        window?.isUserInteractionEnabled = false
        ProgressHUD.show()
    }

    static func dismiss() {
        isShowing = false
        window?.isUserInteractionEnabled = true
        ProgressHUD.dismiss()
    }
}
```

2. Во viewDidAppear добавить проверку:
```
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)

        guard UIBlockingProgressHUD.isShowing == false else { return }
        if let token = OAuth2TokenStorage.shared.token {
            fetchProfile(with: token)
            switchToTabBarController()
        } else {
            guard let authViewController = UIStoryboard(name: "Main", bundle: .main).instantiateViewController(withIdentifier: "AuthViewController") as? AuthViewController else {
                assertionFailure("Failed to show Authentication Screen")
                return
            }

            authViewController.delegate = self
            authViewController.modalPresentationStyle = .fullScreen
            present(authViewController, animated: true)
        }
    }
```

3. После того как будет показан алерт с ошибкой, на нажатие кнопки ok нужно добавить показ экрана авторизации

Georgy-hook committed:
- Fixed retain cycle
- Fixed force unwrap
- Fixed memory leaks related to Date Formatter
- Fixed StatusBar
- Fixed Backgrounds
- Fixed SingleImageViewController constraints
- Fixed Cell selection style
- Fixed image loading in SingleImageViewController
